rskj-core/src/main/java/org/ethereum/core/BlockTxSignatureCache.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.ethereum.core;
+
+import co.rsk.core.RskAddress;
+import co.rsk.remasc.RemascTransaction;
+import co.rsk.util.MaxSizeHashMap;
+
+import java.util.Map;
+
+public class BlockTxSignatureCache implements SignatureCache {
+
+    private static final int MAX_CACHE_SIZE = 6000;
+
+    private final Map<Transaction, RskAddress> addressesCache;
+    private SignatureCache cacheN2;
+
+    public BlockTxSignatureCache(ReceivedTxSignatureCache cacheN2) {
+        this.cacheN2 = cacheN2;
+        addressesCache = new MaxSizeHashMap<>(MAX_CACHE_SIZE,true);
+    }
+
+    public RskAddress getSender(Transaction transaction) {
+
+        if (transaction instanceof RemascTransaction) {
+            return RemascTransaction.REMASC_ADDRESS;
+        }
+
+        if (containsTx(transaction)) {
+            return addressesCache.get(transaction);
+        }
+
+        if (cacheN2.containsTx(transaction)) {
+            RskAddress sender = cacheN2.getSender(transaction);
+            addressesCache.put(transaction, sender);
+            return sender;
+        }
+
+        return transaction.getSender();
+    }
+
+    @Override
+    public boolean containsTx(Transaction transaction) {
+        return addressesCache.containsKey(transaction);
+    }
+
+    @Override
+    public void storeSender(Transaction transaction) {
+
+        if (transaction instanceof RemascTransaction) {
+            return;
+        }
+
+        if (cacheN2.containsTx(transaction)) {
+            RskAddress sender = cacheN2.getSender(transaction);
```

Comment:
The problem with that solution is `getSender` method has two chances
1. If sender is in cache, it returns the queried sender
2. If it is not, the cache computes sender from the signature and then returns it. 

Then in the case which BlockTxSignatureCache queries a transaction sender to cacheN2, it first needs to verify if cacheN2 containsTx and in affirmative case, compute cacheN2.getSender. Because in the opposite case, cacheN2 will extract sender from the signature (which is not desired)

Reply:
What i mean is instead of 
if (cacheN2.containsTx(transaction)) {
            RskAddress sender = cacheN2.getSender(transaction);
            addressesCache.put(transaction, sender);
}
Use
RskAddress sender = cacheN2.getSender(transaction);
if (sender != null) {
            addressesCache.put(transaction, sender);
}

So you have only 1 access